### ISSUE TITLE
Release pipeline does not work with Java 17

### ISSUE THREAD

**Issue by wborn** [openhab/openhab-distro#1445]
The release pipeline job uses the unleash-maven-plugin which is not compatible with Java 17 (see also #1444).

It logs the following error in the [sandbox-openhab4-release](https://ci.openhab.org/view/Sandbox/job/sandbox-openhab4-release/) job:

`[ERROR] Failed to execute goal com.itemis.maven.plugins:unleash-maven-plugin:2.9.3:perform (default-cli) on project org.openhab.core.reactor: Execution default-cli of goal com.itemis.maven.plugins:unleash-maven-plugin:2.9.3:perform failed: WELD-001524: Unable to load proxy class for bean Implicit Bean [javax.enterprise.event.Event] with qualifiers [@Default] with class interface javax.enterprise.event.Event using classloader ClassRealm[plugin>com.itemis.maven.plugins:unleash-maven-plugin:2.9.3, parent: jdk.internal.loader.ClassLoaders$AppClassLoader@5cb0d902]: Could not initialize class org.jboss.classfilewriter.ClassFile -> [Help 1]`

The [unleash-maven-plugin](https://github.com/shillner/unleash-maven-plugin) is no longer maintained so if we want to keep using it we would need to:
* fork it and keep maintaining it ourselves
* upgrade Weld to at least 3.1.1 which fixes this issue ([WELD-2572](https://issues.redhat.com/browse/WELD-2572), [fix](https://github.com/weld/core/pull/1910), [announcement](http://weld.cdi-spec.org/news/2019/05/07/weld-311Final/))

The other alternative is to replace the unleash-maven-plugin with the [maven-release-plugin](https://maven.apache.org/maven-release/maven-release-plugin/) but that could increase release times and be more error prone.

It will be difficult to make milestone/final releases without a working release pipeline.

**Comment by wborn** [comment]
Did you already find some time to look into this issue @pfink?

**Comment by pfink** [comment]
Yes.  But I only tried the unleash fork option yet, because I thought it would be the faster fix. I could fix the exception by upgrading the unleash plugin to Weld 3.1.6 which is possible with a small change in the source code. Anyhow, the next exception is coming:

```
[ERROR] Failed to execute goal com.itemis.maven.plugins:unleash-maven-plugin:2.10.1-SNAPSHOT:perform (default-cli) on project org.openhab.core.reactor: Execution default-cli of goal com.itemis.maven.plugins:unleash-maven-plugin:2.10.1-SNAPSHOT:perform failed: Unable to make protected final java.lang.Class java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int) throws java.lang.ClassFormatError accessible: module java.base does not "opens java.lang" to unnamed module @6ad6ae45 -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal com.itemis.maven.plugins:unleash-maven-plugin:2.10.1-SNAPSHOT:perform (default-cli) on project org.openhab.core.reactor: Execution default-cli of goal com.itemis.maven.plugins:unleash-maven-plugin:2.10.1-SNAPSHOT:perform failed: Unable to make protected final java.lang.Class java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int) throws java.lang.ClassFormatError accessible: module java.base does not "opens java.lang" to unnamed module @6ad6ae45
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:215)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:957)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:289)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:193)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:77)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:568)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
Caused by: org.apache.maven.plugin.PluginExecutionException: Execution default-cli of goal com.itemis.maven.plugins:unleash-maven-plugin:2.10.1-SNAPSHOT:perform failed: Unable to make protected final java.lang.Class java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int) throws java.lang.ClassFormatError accessible: module java.base does not "opens java.lang" to unnamed module @6ad6ae45
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:148)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:957)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:289)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:193)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:77)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:568)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
Caused by: java.lang.reflect.InaccessibleObjectException: Unable to make protected final java.lang.Class java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int) throws java.lang.ClassFormatError accessible: module java.base does not "opens java.lang" to unnamed module @6ad6ae45
    at java.lang.reflect.AccessibleObject.checkCanSetAccessible (AccessibleObject.java:354)
    at java.lang.reflect.AccessibleObject.checkCanSetAccessible (AccessibleObject.java:297)
    at java.lang.reflect.Method.checkCanSetAccessible (Method.java:199)
    at java.lang.reflect.Method.setAccessible (Method.java:193)
    at org.jboss.weld.util.bytecode.ClassFileUtils$1.run (ClassFileUtils.java:67)
    at java.security.AccessController.doPrivileged (AccessController.java:569)
    at org.jboss.weld.util.bytecode.ClassFileUtils.makeClassLoaderMethodsAccessible (ClassFileUtils.java:60)
    at org.jboss.weld.bootstrap.WeldStartup.startContainer (WeldStartup.java:220)
    at org.jboss.weld.bootstrap.WeldBootstrap.startContainer (WeldBootstrap.java:72)
    at org.jboss.weld.environment.se.Weld.initialize (Weld.java:808)
    at com.itemis.maven.plugins.cdi.AbstractCDIMojo.execute (AbstractCDIMojo.java:196)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:957)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:289)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:193)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:77)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:568)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
```

**Comment by pfink** [comment]
Probably, using the maven release plugin would be the better option for the long-term, but for now I guess it would also require significant efforts.

Another idea could be to run the build step with Java 17, and the unleash steps with Java 11 as a fast fix to get the pipeline running again. This could give us some time for migrating to the maven release plugin. I could give that a try this week.

**Comment by wborn** [comment]
Thanks for looking into this!

> Caused by: java.lang.reflect.InaccessibleObjectException: Unable to make protected final java.lang.Class java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int) throws java.lang.ClassFormatError accessible: module java.base does not "opens java.lang" to unnamed module @6ad6ae45

That looks like [WELD-2620](https://issues.redhat.com/browse/WELD-2620) which might be fixed in even newer Weld versions. 

You could also try to work around it by opening the package using some additional MAVEN_OPTS:

`MAVEN_OPTS=--add-opens java.base/java.lang=ALL-UNNAMED`

**Comment by pfink** [comment]
Thanks! I'm on travel until Friday, will give it a try on the weekend.

**Comment by pfink** [comment]
> `MAVEN_OPTS=--add-opens java.base/java.lang=ALL-UNNAMED`

Thanks, with this it seems to run fine now, even though I didn't test the whole release chain. Remark: It'll fail with Maven versions older than 3.8.3 because they contain CDI API 1.0 which conflicts with 2.0 required by Weld 3. As of 3.8.3, Maven does only contain the Inject API.

I pushed my changes to my own forks already, so how do we continue? I don't have the permissions to fork the projects to openHAB and I'm not sure if the openHAB project wants to maintain it. I could either keep it in my own repo and publish it to Maven Central, or we move it to the openHAB project and set me as a maintainer, both would be fine for me.


These would be the projects to fork:

https://github.com/shillner/unleash-maven-plugin
https://github.com/shillner/unleash-scm-provider-git
https://github.com/shillner/maven-cdi-plugin-utils

**Comment by kaikreuzer** [comment]
@pfink Any update here? We have reached February and it would be time for a first milestone release of openHAB 4...

**Comment by kaikreuzer** [comment]
@pfink Would be great to get an update from you!

**Comment by hmerk** [comment]
So we missed one milestone completely due to this issue, what a pitty☹️
@pfink Can we have an update please….

**Comment by pfink** [comment]
I'm sorry for the delay due to me, I was ill the last weeks :( Will continue asap on this.

**Comment by hmerk** [comment]
Sorry to hear @pfink !!! Hope it was not too serious and you are feeling better now.

**Comment by wborn** [comment]
> We have reached February and it would be time for a first milestone release of openHAB 4

To prevent compatibility issues between milestones the Karaf upgrade PRs (#1446) also need to be merged before the first milestone can be made.

**Comment by pfink** [comment]
Update: On the upcoming weekend, there will be siginificant progress on this, I'm working on it.

**Comment by pfink** [comment]
I think the new plugin version should be ready now, but it's a bit hard to test the whole pipeline right now because there are already one or more issues that were introduced. Anyhow, I don't want to go through the whole process of releasing the plugin before knowing that there are no remaining issues with it.

So I tried the quick-fix I suggested earlier

> Another idea could be to run the build step with Java 17, and the unleash steps with Java 11 as a fast fix to get the pipeline running again.

and it seems to work fine with that.

4 PRs / issues I opened:

- #1474 -> This quick-fixes this issue for now and allows us to fix all other issues at the release pipeline. Not a beautiful fix, but it works and I think it can be used to create production releases without any worries.
- https://github.com/openhab/openhab-core/issues/3420 -> issue that break openhab-core release
- https://github.com/openhab/infrastructure/pull/16 -> prepare for testing newer versions of the unleash plugin
- https://github.com/openhab/openhab-distro/issues/1475 -> as soon as #1474 is merged, I think we should close this issue and follow up on that one

**Comment by wborn** [comment]
The pipeline is succeeding again! 🥳 
Now we can start planning the first milestone build.

**Comment by kaikreuzer** [comment]
Hooray! 🎉 
Many thanks, @pfink - great work!

**Comment by kaikreuzer** [comment]
It does not seem to be fully stable yet, though, see https://ci.openhab.org/job/sandbox-openhab4-release/850/
This failed on Z-Wave, while there wasn't any change in the repo and the SNAPSHOT references should have been replaced by the pipeline:
```
[ERROR] 08:19:29,872  	The following modules have references to SNAPSHOT parents that are not scheduled for release:
[ERROR] 08:19:29,872  		org.openhab.addons.bundles:org.openhab.binding.zwave:4.0.0-SNAPSHOT => org.openhab.addons:org.openhab.addons.reactor:4.0.0-SNAPSHOT
[ERROR] An exception was caught while processing the workflow step with id 'checkParentVersions'.
```

**Comment by wborn** [comment]
The sandbox-openhab3-patch-release was running concurrently. I don't know if both jobs cleanup/deploy to the same Maven repos causing issues? I've launched another OH4 sandbox build without an OH3 sandbox build running.

**Comment by kaikreuzer** [comment]
That might indeed be a reason - let's see if the new job succeeds and we can close this issue again.

**Comment by hmerk** [comment]
Zwave succedded ;-)

**Comment by pfink** [comment]
> I don't know if both jobs cleanup/deploy to the same Maven repos causing issues?

Definitely that's the issue. I installed and configured the [Throttle Concurrent Builds Plugin](https://plugins.jenkins.io/throttle-concurrents/) to prevent this in the future.

Just to document what I did:

- install the plugin
- configure a category "sandbox-release" in Jenkins System Configuration and set Total Number of Concurrent Build to `1`
- assign sandbox-openhab3-patch-release and sandbox-openhab4-release to this category in their job configuration

I think it should work, but I didn't want to test right now because I don't want to risk breaking the running build.
